src/nav_channel_ctrl.cpp

```diff
@@ -0,0 +1,259 @@
+#include <nav_msgs/Odometry.h>
+#include <cmath>
+#include <iostream>
+
+class NavChannel {
+public:
+    NavChannel(): nh_(""), private_nh_("~")
+    {
+        // ROS parameters
+        private_nh_.param<double>("error", error, 1.0); // leniency for if we are at the gate
+        private_nh_.param<double>("gate_max_dist", gate_max_dist, 8.0); // max distance for gate 1, determines if we are looking at the right gate, 8.0m is roughly 25ft, can make more accurate if needed.
+        private_nh_.param<double>("gate_max_width", gate_max_width, 4.0); // max distance between two buoys for a gate, confirming we have the two correct ones. 4.0m is roughly 13ft, accounts for max gap and diameter.
+
+
+        // ROS subscribers
+
+        // task_to_exec will confirm we are on nav_channel, then execute task
+        task_to_exec_ = nh_.subscribe("task_to_execute", 10, &NavChannel::navChannelCallback, this);
+        
+        // props will get the prop map and store it
+        prop_map_ = nh_.subscribe("props", 10, &NavChannel::propMapCallback, this);
+
+        // global_position/local grabs position from mavros and stores it
+        global_pos_ = nh_.subscribe("mavros/global_position/local", 10, &NavChannel::globalPositionCallback, this);
+
+
+        // ROS publishers
+
+        // we publish to task_status to update us on progress of task
+        task_status_ = nh_.advertise<nav_channel::TaskStatus>("task_status", 10);
+
+        // we publish to task_goal_position to send the boat to a waypoint
+        task_goal_position_ = nh_.advertise<nav_channel::TaskGoalPosition>("task_goal_position", 10);
+    }
+
+    void spin() {
+        ros::Rate rate(10);
+        while (ros::ok()) {
+            ros::spinOnce();
+            rate.sleep();
+        }
+    }
+
+    void setDestination(geometry_msgs::PoseStamped midpoint) {
+        // sets goal_pos and then publishes it
+        goal_pos.goal_pose = midpoint;
+        // does this set orientation to (0,0,0,0), do we need to grab current orientation and set that to new orientation?
+
+        ROS_INFO_STREAM("Midpoint set at " << midpoint.pose.position.x << ","<< midpoint.pose.position.y << "," << midpoint.pose.position.z);
+
+        task_goal_position_.publish(goal_pos);
+    }
+
+    geometry_msgs::PoseStamped findMidpoint(int gate) {
+        // validates positions of props, then calculates midpoint, and returns
+        // it as PoseStamped
+        geometry_msgs::PoseStamped midpoint;
+
+        bool red = false;   // use two booleans to determine if props exist
+        bool green = false;
+        prop_mapper::Prop red_prop;
+        prop_mapper::Prop green_prop;
+        ROS_DEBUG_STREAM("Gate #" << gate);
+
+        for (int i = 0; (!red || !green) || i > sizeof(props.props) ; i++) {
+            double dist_to_gate = sqrt(pow(props.props[i].vector.x - current_pos.pose.pose.position.x, 2) + pow(props.props[i].vector.y - current_pos.pose.pose.position.y, 2));
```

Comment:
would it make more sense to do this distance check after we've identified that we have a red and green marker? There could be lots of other props in the array.